File: phenotype-mcp-sdk-go/transport_test.go
package mcpsdk

import (
	"bytes"
	"context"
	"encoding/json"
	stderrors "errors"
	"strings"
	"testing"
)

// mockServer is the Go equivalent of the Rust `MockServer` test fixture. It
// exposes one tool ("echo") and a read_resource method that always reports
// the URI as missing, mirroring the Rust implementation verbatim.
type mockServer struct {
	tools     []Tool
	resources []Resource
}

func newMockServer() *mockServer {
	return &mockServer{
		tools: []Tool{NewTool(
			"echo",
			"echo",
			json.RawMessage(`{}`),
			func(_ context.Context, args json.RawMessage) (any, error) {
				return args, nil
			},
		)},
	}
}

func (m *mockServer) ListTools() []Tool         { return m.tools }
func (m *mockServer) ListResources() []Resource { return m.resources }
func (m *mockServer) CallTool(_ context.Context, _ string, _ json.RawMessage) (any, error) {
	return "ok", nil
}
func (m *mockServer) ReadResource(_ context.Context, uri string) (ResourceContent, error) {
	return ResourceContent{}, &McpError{Kind: KindResourceNotFound, Message: uri}
}

// TestStdioTransportNew mirrors the Rust test `stdio_transport_new`: a
// freshly constructed transport must have the server wired up and report
// its tool count correctly.
func TestStdioTransportNew(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	if got := len(tr.Server().ListTools()); got != 1 {
		t.Fatalf("ListTools len: got %d, want 1", got)
	}
}

// TestSseTransportNew mirrors the Rust test `sse_transport_new`: a freshly
// constructed SseTransport must accept a serve call and return nil (the
// implementation is currently a stub).
func TestSseTransportNew(t *testing.T) {
	t.Helper()
	tr := NewSseTransport(newMockServer())
	if err := tr.Serve("127.0.0.1:0"); err != nil {
		t.Fatalf("Serve: %v", err)
	}
}

// TestSseTransportServerAccess mirrors the Rust test
// `sse_transport_server_access`: the transport must hand back the inner
// server so embedders can introspect it.
func TestSseTransportServerAccess(t *testing.T) {
	t.Helper()
	tr := NewSseTransport(newMockServer())
	tools := tr.Server().ListTools()
	if len(tools) != 1 {
		t.Fatalf("tools len: got %d, want 1", len(tools))
	}
	if tools[0].Name != "echo" {
		t.Fatalf("tool name: got %q, want %q", tools[0].Name, "echo")
	}
}

// TestStdioTransportStringDoesNotPanic mirrors the Rust test
// `stdio_transport_debug`: the transport must be safely representable as
// a string (we use fmt.Stringer-equivalent String() instead of fmt.Sprintf
// %v, but the property — "does not panic" — is the same).
func TestStdioTransportStringDoesNotPanic(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	_ = tr // the transport is a thin holder; the test is that we can build one
}

// TestStdioTransportHandleToolsList mirrors the Rust test
// `stdio_transport_handle_tools_list`: a `tools/list` request must yield a
// response containing a `tools` key.
func TestStdioTransportHandleToolsList(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	resp, err := tr.handleRequest(context.Background(), []byte(`{"method":"tools/list"}`))
	if err != nil {
		t.Fatalf("handleRequest: %v", err)
	}
	if !hasKey(t, resp, "tools") {
		t.Fatalf("expected `tools` key in response: %s", resp)
	}
}

// TestStdioTransportHandleUnknownMethod mirrors the Rust test
// `stdio_transport_handle_unknown_method`: a bogus method must yield an
// error whose message contains "unknown method".
func TestStdioTransportHandleUnknownMethod(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	resp, err := tr.handleRequest(context.Background(), []byte(`{"method":"bogus"}`))
	if err == nil {
		t.Fatalf("expected error, got response %s", resp)
	}
	if !strings.Contains(err.Error(), "unknown method") {
		t.Fatalf("error message %q does not contain %q", err.Error(), "unknown method")
	}
}

// TestStdioTransportHandleToolsCall mirrors the Rust test
// `stdio_transport_handle_tools_call`: a well-formed `tools/call` request
// must yield a response containing a `result` key.
func TestStdioTransportHandleToolsCall(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	resp, err := tr.handleRequest(context.Background(), []byte(`{"method":"tools/call","params":{"name":"echo","arguments":{}}}`))
	if err != nil {
		t.Fatalf("handleRequest: %v", err)
	}
	if !hasKey(t, resp, "result") {
		t.Fatalf("expected `result` key in response: %s", resp)
	}
}

// TestStdioTransportHandleResourcesList is an extra coverage test: an
// empty `resources/list` request must yield a response containing a
// `resources` key (even when the list is empty).
func TestStdioTransportHandleResourcesList(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	resp, err := tr.handleRequest(context.Background(), []byte(`{"method":"resources/list"}`))
	if err != nil {
		t.Fatalf("handleRequest: %v", err)
	}
	if !hasKey(t, resp, "resources") {
		t.Fatalf("expected `resources` key in response: %s", resp)
	}
}

// TestStdioTransportHandleResourcesRead covers the fourth method dispatched
// by the transport: `resources/read`. The mock server always reports
// ResourceNotFound, so we expect a *McpError of the matching kind.
func TestStdioTransportHandleResourcesRead(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	resp, err := tr.handleRequest(context.Background(), []byte(`{"method":"resources/read","params":{"uri":"res://missing"}}`))
	if err == nil {
		t.Fatalf("expected error, got %s", resp)
	}
	var me *McpError
	if !errorsAs(err, &me) || me.Kind != KindResourceNotFound {
		t.Fatalf("expected *McpError{Kind: ResourceNotFound}, got %v", err)
	}
}

// TestStdioTransportHandleInvalidJSON covers the JSON-decode failure path.
func TestStdioTransportHandleInvalidJSON(t *testing.T) {
	t.Helper()
	tr := NewStdioTransport(newMockServer())
	resp, err := tr.handleRequest(context.Background(), []byte(`{not json`))
	if err == nil {
		t.Fatalf("expected error, got %s", resp)
	}
	var me *McpError
	if !errorsAs(err, &me) || me.Kind != KindInvalidArguments {
		t.Fatalf("expected *McpError{Kind: InvalidArguments}, got %v", err)
	}
}

// TestStdioTransportRunWith exercises the full stdin/stdout dispatch loop
// against an in-memory pipe. The server sees three requests and the
// reader collects three newline-delimited JSON responses.
func TestStdioTransportRunWith(t *testing.T) {
	t.Helper()

	in := bytes.NewBufferString(strings.Join([]string{
		`{"method":"tools/list"}`,
		`{"method":"tools/call","params":{"name":"echo","arguments":{"x":1}}}`,
		`{"method":"resources/list"}`,
		"", // empty line — should be ignored, no response emitted
		`{"method":"bogus"}`, // produces no response (handleRequest returns an error envelope in RunWith)
	}, "\n") + "\n")

	var out bytes.Buffer
	tr := NewStdioTransport(newMockServer())
	if err := tr.RunWith(context.Background(), in, &out); err != nil {
		t.Fatalf("RunWith: %v", err)
	}

	lines := splitNonEmptyLines(out.String())
	if len(lines) != 4 {
		t.Fatalf("expected 4 response lines (the empty input is skipped), got %d: %q", len(lines), out.String())
	}
	// First response: tools/list
	if !hasKey(t, json.RawMessage(lines[0]), "tools") {
		t.Fatalf("line 0: expected `tools`, got %s", lines[0])
	}
	// Second response: tools/call
	if !hasKey(t, json.RawMessage(lines[1]), "result") {
		t.Fatalf("line 1: expected `result`, got %s", lines[1])
	}
	// Third response: resources/list
	if !hasKey(t, json.RawMessage(lines[2]), "resources") {
		t.Fatalf("line 2: expected `resources`, got %s", lines[2])
	}
	// Fourth response: error envelope for the bogus method
	if !hasKey(t, json.RawMessage(lines[3]), "error") {
		t.Fatalf("line 3: expected `error` envelope, got %s", lines[3])
	}
}

// TestMcpServerBaseEmbed exercises the embeddable McpServerBase helper.
func TestMcpServerBaseEmbed(t *testing.T) {
	t.Helper()

	type srv struct {
		McpServerBase
	}
	tool := NewTool("hi", "Say hi", json.RawMessage(`{}`), func(_ context.Context, _ json.RawMessage) (any, error) {
		return "hello", nil
	})
	s := &srv{McpServerBase: McpServerBase{Tools: []Tool{tool}}}

	if got, err := s.CallTool(context.Background(), "hi", nil); err != nil {
		t.Fatalf("CallTool: %v", err)
	} else if got != "hello" {
		t.Fatalf("CallTool result: got %v, want hello", got)
	}

	if _, err := s.CallTool(context.Background(), "missing", nil); err == nil {
		t.Fatalf("expected ToolNotFound error, got nil")
	} else {
		var me *McpError
		if !errorsAs(err, &me) || me.Kind != KindToolNotFound {
			t.Fatalf("expected *McpError{Kind: ToolNotFound}, got %v", err)
		}
	}

	if _, err := s.ReadResource(context.Background(), "res://x"); err == nil {
		t.Fatalf("expected ResourceNotFound error, got nil")
	} else {
		var me *McpError
		if !errorsAs(err, &me) || me.Kind != KindResourceNotFound {
			t.Fatalf("expected *McpError{Kind: ResourceNotFound}, got %v", err)
		}
	}
}

// TestMCPResourceFactory is a table-driven test for the MCPResource
// factory's two forms (4-arg full and 2-arg minimal).
func TestMCPResourceFactory(t *testing.T) {
	t.Helper()

	full := MCPResource("res://docs", "Documentation", "text/markdown", "API docs")
	if full.URI != "res://docs" || full.Name != "Documentation" {
		t.Fatalf("full: %+v", full)
	}
	if full.MimeType == nil || *full.MimeType != "text/markdown" {
		t.Fatalf("full.MimeType: %v", full.MimeType)
	}
	if full.Description == nil || *full.Description != "API docs" {
		t.Fatalf("full.Description: %v", full.Description)
	}

	short := MCPResource("res://readme", "Readme")
	if short.URI != "res://readme" || short.Name != "Readme" {
		t.Fatalf("short: %+v", short)
	}
	if short.MimeType != nil {
		t.Fatalf("short.MimeType: expected nil, got %v", *short.MimeType)
	}
	if short.Description != nil {
		t.Fatalf("short.Description: expected nil, got %v", *short.Description)
	}
}

// --- helpers --------------------------------------------------------------

// hasKey checks that the given JSON object contains the given top-level key.
func hasKey(t *testing.T, raw json.RawMessage, key string) bool {
	t.Helper()
	var m map[string]json.RawMessage
	if err := json.Unmarshal(raw, &m); err != nil {
		t.Fatalf("hasKey(%q): unmarshal: %v", key, err)
	}
	_, ok := m[key]
	return ok
}

// errorsAs is a tiny wrapper around errors.As that the rest of the file
// reuses without needing to import the "errors" package in two places.
func errorsAs(err error, target any) bool {
	type wrapper interface{ As(any) bool }
	if w, ok := err.(wrapper); ok {
		return w.As(target)
	}
	// Fall back to the standard library. We import it lazily via a
	// function variable to keep the import block tiny.
	return stderrors.As(err, target)
}

func splitNonEmptyLines(s string) []string {
	var out []string
	for _, line := range strings.Split(s, "\n") {
		if line == "" {
			continue
		}
		out = append(out, line)
	}
	return out
}

File: phenotype-mcp-sdk-ts/test/transport.test.ts
/**
 * Tests for `StdioTransport` and `SseTransport`.
 *
 * Mirrors `phenotype-mcp-sdk-rs` `transport.rs:132-213`:
 *   stdio_transport_new, sse_transport_new, sse_transport_server_access,
 *   stdio_transport_debug, stdio_transport_handle_tools_list,
 *   stdio_transport_handle_unknown_method, stdio_transport_handle_tools_call
 *
 * Also covers the `resources/list` and `resources/read` request shapes
 * (transport.rs:85-97) for parity with the Rust handle_request switch.
 */

import { describe, test } from 'node:test';
import assert from 'node:assert/strict';
import { z } from 'zod';

import {
  StdioTransport,
  SseTransport,
  McpServerImpl,
  McpError,
  mcpTool,
  mcpResource,
  type McpServer,
} from '../src/index.js';

/** Mock server mirroring `MockServer` in transport.rs:139-160. */
class MockServer extends McpServerImpl {
  constructor() {
    super();
    this.addTool(
      mcpTool('echo', 'echo', z.object({}).passthrough(), (args: unknown) => args),
    );
    this.addResource(
      mcpResource('res://readme', 'Readme', 'text/markdown', 'Project readme'),
      { uri: 'res://readme', mimeType: 'text/markdown', text: '# Hello' },
    );
  }

  // Override to mirror the Rust mock: always returns "ok" for any
  // tool name unless the tool exists in the map (in which case the
  // default McpServerImpl path handles it).
  async callTool(name: string, args: unknown): Promise<unknown> {
    if (name === 'special') return 'mock-special';
    return super.callTool(name, args);
  }

  // Override to mirror the Rust mock: read_resource returns
  // ResourceNotFound for any uri unless it's the registered one.
  async readResource(uri: string) {
    if (uri === 'res://readme') {
      return { uri, mimeType: 'text/markdown', text: '# Hello' };
    }
    throw McpError.resourceNotFound(uri);
  }
}

describe('StdioTransport', () => {
  test('stdio_transport_new', () => {
    // Mirrors transport.rs:163-166
    const server = new MockServer();
    const transport = new StdioTransport(server);
    assert.equal(transport.getServer(), server);
  });

  test('stdio_transport_to_string', () => {
    // Mirrors transport.rs:183-187 stdio_transport_debug —
    // verifies that stringification doesn't panic and includes the
    // class name.
    const transport = new StdioTransport(new MockServer());
    const s = transport.toString();
    assert.ok(s.includes('StdioTransport'));
  });

  test('stdio_transport_handle_tools_list', async () => {
    // Mirrors transport.rs:190-195
    const transport = new StdioTransport(new MockServer());
    const resp = (await transport.handleRequest({ method: 'tools/list' })) as {
      tools: Array<{ name: string; description: string; inputSchema: unknown }>;
    };
    assert.ok(Array.isArray(resp.tools));
    assert.ok(resp.tools.length >= 1);
    const echo = resp.tools.find((t) => t.name === 'echo');
    assert.ok(echo);
    assert.equal(echo.description, 'echo');
    assert.ok(echo.inputSchema);
  });

  test('stdio_transport_handle_unknown_method', async () => {
    // Mirrors transport.rs:198-204
    const transport = new StdioTransport(new MockServer());
    await assert.rejects(
      async () => transport.handleRequest({ method: 'bogus' }),
      (err: unknown) =>
        err instanceof McpError &&
        err.kind === 'InvalidArguments' &&
        err.detail.includes('unknown method'),
    );
  });

  test('stdio_transport_handle_tools_call', async () => {
    // Mirrors transport.rs:207-212
    const transport = new StdioTransport(new MockServer());
    const resp = (await transport.handleRequest({
      method: 'tools/call',
      params: { name: 'echo', arguments: {} },
    })) as { result: unknown };
    assert.ok('result' in resp);
  });

  test('stdio_transport_handle_tools_call_missing_name', async () => {
    // TS-specific: validate that params.name is required.
    const transport = new StdioTransport(new MockServer());
    await assert.rejects(
      async () => transport.handleRequest({ method: 'tools/call', params: {} }),
      (err: unknown) =>
        err instanceof McpError && err.kind === 'InvalidArguments',
    );
  });

  test('stdio_transport_handle_resources_list', async () => {
    // Mirrors transport.rs:85-88
    const transport = new StdioTransport(new MockServer());
    const resp = (await transport.handleRequest({ method: 'resources/list' })) as {
      resources: Array<{ uri: string; name: string }>;
    };
    assert.ok(Array.isArray(resp.resources));
    assert.equal(resp.resources.length, 1);
    assert.equal(resp.resources[0].uri, 'res://readme');
  });

  test('stdio_transport_handle_resources_read', async () => {
    // Mirrors transport.rs:89-97
    const transport = new StdioTransport(new MockServer());
    const resp = (await transport.handleRequest({
      method: 'resources/read',
      params: { uri: 'res://readme' },
    })) as { content: { text: string } };
    assert.equal(resp.content.text, '# Hello');
  });

  test('stdio_transport_handle_resources_read_not_found', async () => {
    const transport = new StdioTransport(new MockServer());
    await assert.rejects(
      async () =>
        transport.handleRequest({
          method: 'resources/read',
          params: { uri: 'res://missing' },
        }),
      (err: unknown) =>
        err instanceof McpError && err.kind === 'ResourceNotFound',
    );
  });
});

describe('SseTransport', () => {
  test('sse_transport_new', async () => {
    // Mirrors transport.rs:169-172
    const server: McpServer = new MockServer();
    const transport = new SseTransport(server);
    // serve() is a stub and returns successfully.
    const result = await transport.serve('127.0.0.1:0');
    assert.equal(result, undefined);
  });

  test('sse_transport_server_access', async () => {
    // Mirrors transport.rs:175-180
    const server: McpServer = new MockServer();
    const transport = new SseTransport(server);
    const inner = transport.getServer();
    const tools = await inner.listTools();
    assert.equal(tools.length, 1);
    assert.equal(tools[0].name, 'echo');
  });
});
